pkg/registrar/transitgateway.go

```diff
@@ -73,14 +75,106 @@ func (r *TransitGateway) Register(ctx context.Context, cluster *capi.Cluster) er
+				if err != nil {
+					logger.Error(err, "Failed to get management cluster")
+					return err
+				}
+
+				gatewayID = annotations.GetNetworkTopologyTransitGatewayID(mc)
+				if gatewayID == "" {
+					err = fmt.Errorf("management cluster doesn't have a TGW specified")
+					logger.Error(err, "The Management Cluster doesn't have a Transit Gateway ID specified")
+					return err
+				}
+			}
+
+			tgw, err = r.getTransitGateway(ctx, gatewayID)
+			if err != nil {
+				return err
+			}
+			if tgw == nil {
+				err = fmt.Errorf("failed to find TransitGateway for provided ID")
+				logger.Error(err, "No TransitGateway found for ID provided on annotations", "transitGatewayID", gatewayID)
+				return err
+			}
+		}
+
+		// Ensure TGW ID is saved back to the current cluster
+		baseCluster := cluster.DeepCopy()
+		annotations.SetNetworkTopologyTransitGatewayID(cluster, *tgw.TransitGatewayId)
+		if cluster, err = r.clusterClient.Patch(ctx, cluster, client.MergeFrom(baseCluster)); err != nil {
+			logger.Error(err, "Failed to patch cluster resource with TGW ID")
+			return err
+		}
+
+		awsCluster, err := r.getAWSCluster(ctx, cluster)
+		if err != nil {
+			logger.Error(err, "Failed to get AWSCluster for Cluster")
+			return err
+		}
+
+		var tgwAttachment *types.TransitGatewayVpcAttachment
+		if awsCluster.Spec.NetworkSpec.VPC.ID == "" {
+			logger.Info("vpc not yet ready, skipping attachment for now", "transitGatewayID", tgw.TransitGatewayId)
+			return &VPCNotReadyError{}
+		} else if tgw.State == types.TransitGatewayStateAvailable {
+			tgwAttachment, err = r.attachTransitGateway(ctx, tgw.TransitGatewayId, awsCluster)
+			if err != nil {
+				return err
+			}
+		} else {
+			logger.Info("transit gateway not available, skipping attachment for now", "transitGatewayID", tgw.TransitGatewayId, "tgwState", tgw.State)
+			return &TransitGatewayNotAvailableError{}
+		}
+
+		if tgwAttachment.State == types.TransitGatewayAttachmentStateInitiating || tgwAttachment.State == types.TransitGatewayAttachmentStateInitiatingRequest ||
+			tgwAttachment.State == types.TransitGatewayAttachmentStatePending || tgwAttachment.State == types.TransitGatewayAttachmentStatePendingAcceptance {
+			logger.Info("Sending SNS message")
+
+			_, err = r.transitGatewayClient.PublishSNSMessage(ctx, &sns.PublishInput{
+				Message: aws.String("Request TransitGatewayAttachment"),
+				MessageAttributes: map[string]snstypes.MessageAttributeValue{
+					"Postfach":      {DataType: aws.String("String"), StringValue: aws.String("")},
+					"Account_ID":    {DataType: aws.String("String"), StringValue: tgwAttachment.VpcOwnerId},
+					"Attachment_ID": {DataType: aws.String("String"), StringValue: tgwAttachment.TransitGatewayAttachmentId},
+					"CIDR":          {DataType: aws.String("String"), StringValue: &awsCluster.Spec.NetworkSpec.VPC.CidrBlock},
+					"Name":          {DataType: aws.String("String"), StringValue: &cluster.ObjectMeta.Name},
+				},
+			})
+			if err != nil {
```

Comment:
if the SNS message fail to send(network or other issue), will the next reconciliation loop try to send it again? or will it fail on the line  https://github.com/giantswarm/aws-network-topology-operator/pull/60/files#diff-0b7fb972fc7f67d673bb22e706aae67bc06d1bf8d7e9e52cfb1ac5b8a7e495c5R146 where it tries to attach it again ( and since the operation is already in some `pending` state it will fail)

Reply:
When creating the TGW _attachment_ it first checks for an existing one 

https://github.com/giantswarm/aws-network-topology-operator/blob/324f1f04ba895b0dc20d55d80923cd20a2645f41/pkg/registrar/transitgateway.go#L426

If found it uses that and then based on the state of that TGW will either decide to try and send the SNS message again or not bother as the attachment has already been processed.